File: src/parser.js
/**
 * Parser for TradeMinutes DSL
 * Converts tokens into an Abstract Syntax Tree (AST)
 */

const Lexer = require('./lexer');

class Parser {
  constructor(input) {
    this.lexer = new Lexer(input);
    this.tokens = [];
    this.current = 0;
  }

  /**
   * Parse the input and return AST
   * @returns {Object} Abstract Syntax Tree
   */
  parse() {
    this.tokens = this.lexer.tokenize();
    this.current = 0;
    
    const ast = {
      type: 'Program',
      workflows: []
    };

    while (!this.isAtEnd()) {
      if (this.check('KEYWORD', 'workflow')) {
        ast.workflows.push(this.parseWorkflow());
      } else {
        this.advance(); // Skip unknown tokens
      }
    }

    return ast;
  }

  /**
   * Parse a workflow definition
   */
  parseWorkflow() {
    this.consume('KEYWORD', 'workflow', 'Expected "workflow" keyword');
    
    const name = this.consume('STRING', null, 'Expected workflow name');
    this.consume('LBRACE', null, 'Expected "{" after workflow name');
    
    const steps = [];
    while (!this.check('RBRACE') && !this.isAtEnd()) {
      steps.push(this.parseStep());
    }
    
    this.consume('RBRACE', null, 'Expected "}" to close workflow');
    
    return {
      type: 'Workflow',
      name: name.value,
      steps
    };
  }

  /**
   * Parse a step definition
   */
  parseStep() {
    this.consume('KEYWORD', 'step', 'Expected "step" keyword');
    
    const stepNumber = this.consume('NUMBER', null, 'Expected step number');
    this.consume('COLON', null, 'Expected ":" after step number');
    
    const command = this.parseCommand();
    
    return {
      type: 'Step',
      number: stepNumber.value,
      command
    };
  }

  /**
   * Parse a command with its arguments
   */
  parseCommand() {
    const commandName = this.consume('KEYWORD', null, 'Expected command name');
    this.consume('LPAREN', null, 'Expected "(" after command name');
    
    const args = [];
    
    // Parse arguments
    while (true) {
      const currentToken = this.peek();
      
      // If we're at the closing parenthesis, we're done
      if (currentToken.type === 'RPAREN') {
        break;
      }
      
      // If we're at the end, that's an error
      if (currentToken.type === 'EOF') {
        throw new Error(`Unexpected end of input at position ${currentToken.position}`);
      }
      
      // Parse the argument
      args.push(this.parseArgument());
      
      // Check if we're done or need a comma
      const nextToken = this.peek();
      if (nextToken.type === 'RPAREN') {
        break;
      } else if (nextToken.type === 'COMMA') {
        this.advance(); // consume the comma
      } else {
        throw new Error(`Expected "," or ")" at position ${nextToken.position}`);
      }
    }
    
    // Consume the closing parenthesis
    this.consume('RPAREN', null, 'Expected ")" to close command arguments');
    
    return {
      type: 'Command',
      name: commandName.value,
      arguments: args
    };
  }

  /**
   * Parse a command argument
   */
  parseArgument() {
    if (this.match('STRING', null)) {
      return {
        type: 'StringLiteral',
        value: this.previous().value
      };
    } else if (this.match('NUMBER', null)) {
      return {
        type: 'NumberLiteral',
        value: this.previous().value
      };
    } else if (this.match('KEYWORD', 'step')) {
      // Reference to another step
      const stepRef = this.consume('NUMBER', null, 'Expected step number after "step"');
      return {
        type: 'StepReference',
        stepNumber: stepRef.value
      };
    } else if (this.match('IDENTIFIER', null)) {
      return {
        type: 'Identifier',
        value: this.previous().value
      };
    } else {
      throw new Error(`Unexpected token at position ${this.peek().position}`);
    }
  }

  /**
   * Check if current token matches expected type and value
   */
  match(type, value) {
    if (this.check(type, value)) {
      this.advance();
      return true;
    }
    return false;
  }

  /**
   * Check if current token is of expected type and value
   */
  check(type, value) {
    if (this.isAtEnd()) return false;
    const token = this.peek();
    return token.type === type && (value === null || value === undefined || token.value === value);
  }

  /**
   * Consume a token of expected type and value
   */
  consume(type, value, message) {
    if (this.check(type, value)) {
      return this.advance();
    }
    throw new Error(`${message} at position ${this.peek().position}`);
  }

  /**
   * Advance to next token
   */
  advance() {
    if (!this.isAtEnd()) {
      this.current++;
    }
    return this.previous();
  }

  /**
   * Check if we've reached the end of tokens
   */
  isAtEnd() {
    return this.peek().type === 'EOF';
  }

  /**
   * Get current token without advancing
   */
  peek() {
    return this.tokens[this.current];
  }

  /**
   * Get previous token
   */
  previous() {
    return this.tokens[this.current - 1];
  }
}

module.exports = Parser; 

File: rust-version/src/parser.rs
use anyhow::{anyhow, Result};
use crate::ast::*;
use crate::lexer::{Token, TokenType};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, current: 0 }
    }
    
    pub fn parse(&mut self) -> Result<Program> {
        let mut workflows = Vec::new();
        let mut variables = Vec::new();
        
        while !self.is_at_end() {
            match self.peek().token_type {
                TokenType::Workflow => {
                    workflows.push(self.parse_workflow()?);
                }
                TokenType::Let | TokenType::Var | TokenType::Const => {
                    variables.push(self.parse_variable_declaration()?);
                }
                _ => {
                    return Err(anyhow!("Expected workflow or variable declaration"));
                }
            }
        }
        
        Ok(Program { workflows, variables })
    }
    
    fn parse_workflow(&mut self) -> Result<Workflow> {
        self.consume(TokenType::Workflow, "Expected 'workflow'")?;
        
        let name = self.consume_string("Expected workflow name")?;
        
        self.consume(TokenType::LeftBrace, "Expected '{' after workflow name")?;
        
        let mut steps = Vec::new();
        while !self.check(TokenType::RightBrace) && !self.is_at_end() {
            steps.push(self.parse_step()?);
        }
        
        self.consume(TokenType::RightBrace, "Expected '}' after workflow body")?;
        
        Ok(Workflow { name, steps })
    }
    
    fn parse_step(&mut self) -> Result<Step> {
        self.consume(TokenType::Step, "Expected 'step'")?;
        
        let id = self.consume_number("Expected step number")? as u32;
        
        self.consume(TokenType::Colon, "Expected ':' after step number")?;
        
        let content = if self.check(TokenType::If) {
            StepContent::Conditional(self.parse_conditional_statement()?)
        } else {
            StepContent::Command(self.parse_command()?)
        };
        
        Ok(Step { id, content })
    }
    
    fn parse_command(&mut self) -> Result<Command> {
        let name = self.consume_identifier("Expected command name")?;
        
        let arguments = if self.check(TokenType::LeftParen) {
            self.consume(TokenType::LeftParen, "Expected '('")?;
            let args = self.parse_expression_list()?;
            self.consume(TokenType::RightParen, "Expected ')'")?;
            args
        } else {
            Vec::new()
        };
        
        Ok(Command { name, arguments })
    }
    
    fn parse_conditional_statement(&mut self) -> Result<ConditionalStatement> {
        self.consume(TokenType::If, "Expected 'if'")?;
        
        self.consume(TokenType::LeftParen, "Expected '(' after 'if'")?;
        let condition = self.parse_expression()?;
        self.consume(TokenType::RightParen, "Expected ')' after condition")?;
        
        self.consume(TokenType::LeftBrace, "Expected '{' after condition")?;
        let mut if_steps = Vec::new();
        while !self.check(TokenType::RightBrace) && !self.is_at_end() {
            if_steps.push(self.parse_step()?);
        }
        self.consume(TokenType::RightBrace, "Expected '}' after if block")?;
        
        let else_steps = if self.check(TokenType::Else) {
            self.advance(); // consume 'else'
            self.consume(TokenType::LeftBrace, "Expected '{' after 'else'")?;
            let mut steps = Vec::new();
            while !self.check(TokenType::RightBrace) && !self.is_at_end() {
                steps.push(self.parse_step()?);
            }
            self.consume(TokenType::RightBrace, "Expected '}' after else block")?;
            Some(steps)
        } else {
            None
        };
        
        Ok(ConditionalStatement {
            condition,
            if_steps,
            else_steps,
        })
    }
    
    fn parse_variable_declaration(&mut self) -> Result<VariableDeclaration> {
        let keyword = match self.peek().token_type {
            TokenType::Let => "let",
            TokenType::Var => "var",
            TokenType::Const => "const",
            _ => return Err(anyhow!("Expected variable declaration keyword")),
        };
        
        self.advance(); // consume keyword
        
        let name = self.consume_identifier("Expected variable name")?;
        
        self.consume(TokenType::Equal, "Expected '=' after variable name")?;
        
        let value = self.parse_expression()?;
        
        Ok(VariableDeclaration {
            keyword: keyword.to_string(),
            name,
            value,
        })
    }
    
    fn parse_expression(&mut self) -> Result<Expression> {
        self.parse_binary_expression()
    }
    
    fn parse_binary_expression(&mut self) -> Result<Expression> {
        let mut left = self.parse_primary()?;
        
        while self.match_token(&[TokenType::Plus, TokenType::EqualEqual, TokenType::NotEqual, 
                               TokenType::Greater, TokenType::Less, TokenType::GreaterEqual, TokenType::LessEqual]) {
            let operator = self.previous().lexeme.clone();
            let right = self.parse_primary()?;
            left = Expression::binary(left, &operator, right);
        }
        
        Ok(left)
    }
    
    fn parse_primary(&mut self) -> Result<Expression> {
        match self.peek().token_type {
            TokenType::String => {
                let value = self.advance().literal.clone().unwrap_or_default();
                Ok(Expression::string(&value))
            }
            TokenType::Number => {
                let value = self.advance().lexeme.parse::<f64>()
                    .map_err(|_| anyhow!("Invalid number"))?;
                Ok(Expression::number(value))
            }
            TokenType::Identifier => {
                let name = self.advance().lexeme.clone();
                
                // Check for property access (e.g., step 1.status)
                if self.check(TokenType::Dot) {
                    self.advance(); // consume '.'
                    let property = self.consume_identifier("Expected property name")?;
                    Ok(Expression::property_access(Expression::identifier(&name), &property))
                } else {
                    Ok(Expression::identifier(&name))
                }
            }
            TokenType::Step => {
                self.advance(); // consume 'step'
                let step_id = self.consume_number("Expected step number")? as u32;
                
                let property = if self.check(TokenType::Dot) {
                    self.advance(); // consume '.'
                    Some(self.consume_identifier("Expected property name")?)
                } else {
                    None
                };
                
                Ok(Expression::step_reference(step_id, property.as_deref()))
            }
            _ => Err(anyhow!("Expected expression")),
        }
    }
    
    fn parse_expression_list(&mut self) -> Result<Vec<Expression>> {
        let mut expressions = Vec::new();
        
        if !self.check(TokenType::RightParen) {
            loop {
                expressions.push(self.parse_expression()?);
                
                if !self.match_token(&[TokenType::Comma]) {
                    break;
                }
            }
        }
        
        Ok(expressions)
    }
    
    // Helper methods
    fn advance(&mut self) -> &Token {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }
    
    fn check(&self, token_type: TokenType) -> bool {
        if self.is_at_end() {
            false
        } else {
            self.peek().token_type == token_type
        }
    }
    
    fn match_token(&mut self, types: &[TokenType]) -> bool {
        for token_type in types {
            if self.check(token_type.clone()) {
                self.advance();
                return true;
            }
        }
        false
    }
    
    fn consume(&mut self, token_type: TokenType, message: &str) -> Result<&Token> {
        if self.check(token_type) {
            Ok(self.advance())
        } else {
            Err(anyhow!("{}", message))
        }
    }
    
    fn consume_string(&mut self, message: &str) -> Result<String> {
        let token = self.consume(TokenType::String, message)?;
        Ok(token.literal.clone().unwrap_or_default())
    }
    
    fn consume_number(&mut self, message: &str) -> Result<f64> {
        let token = self.consume(TokenType::Number, message)?;
        token.lexeme.parse::<f64>()
            .map_err(|_| anyhow!("{}", message))
    }
    
    fn consume_identifier(&mut self, message: &str) -> Result<String> {
        let token = self.consume(TokenType::Identifier, message)?;
        Ok(token.lexeme.clone())
    }
    
    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }
    
    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }
    
    fn is_at_end(&self) -> bool {
        self.current >= self.tokens.len() || self.peek().token_type == TokenType::Eof
    }
} 